clusterloader2/pkg/measurement/common/api_availability_measurement.go

```diff
@@ -0,0 +1,202 @@
+	}
+}
+
+func createApiAvailabilityMeasurement() measurement.Measurement {
+	return &apiAvailabilityMeasurement{}
+}
+
+type apiAvailabilityOutput struct {
+	AvailabilityPercentage     int `json:"availabilityPercentage"`
+	LongestUnavailableDuration int `json:"longestUnavailableDuration"`
+}
+
+type apiAvailabilityMetrics struct {
+	numProbes                  int
+	numSuccess                 int
+	numFailures                int
+	prevProbeFailure           bool
+	maxConsecutiveFailedProbes int
+	consecutiveFailedProbes    int
+}
+
+type apiAvailabilityMeasurement struct {
+	isRunning              bool
+	stopCh                 chan struct{}
+	hosts                  []string
+	apiAvailabilityMetrics *apiAvailabilityMetrics
+	wg                     sync.WaitGroup
+}
+
+func (a *apiAvailabilityMeasurement) apiServerAvailable(c clientset.Interface, SSHToMasterSupported bool, config *measurement.MeasurementConfig) bool {
+	for _, host := range a.hosts {
+		// Get the api-server measurement data over SSH.
+		// Start by checking that the provider allows us to do so.
+		if !SSHToMasterSupported {
+			_, err := c.CoreV1().RESTClient().Get().AbsPath("/healthz").DoRaw()
+			if err != nil {
+				return false
+			}
+
+			break
+		}
+		command := fmt.Sprintf("curl -s -k %slocalhost:%v/healthz", "https://", 443)
+		_, err := measurementutil.SSH(command, host+":22", "")
+		if err != nil {
+			return false
+		}
+	}
+	return true
+}
+
+func (a *apiAvailabilityMeasurement) start(config *measurement.MeasurementConfig, SSHToMasterSupported bool, probeDuration int) ([]measurement.Summary, error) {
+	a.hosts = config.ClusterFramework.GetClusterConfig().MasterIPs
+	if len(a.hosts) < 1 {
+		return nil, goerrors.Errorf("Api Server measurement can't execute due to no MasterIps")
+	}
+
+	k8sClient := config.ClusterFramework.GetClientSets().GetClient()
+	a.isRunning = true
+	a.stopCh = make(chan struct{})
+	a.wg.Add(1)
+	go func() {
+		defer a.wg.Done()
+		for {
+			select {
+			case <-a.stopCh:
+				return
+			case <-time.After(time.Duration(probeDuration)):
+				apiServerAvailable := a.apiServerAvailable(k8sClient, SSHToMasterSupported, config)
+				// Always update the Probes count
+				a.apiAvailabilityMetrics.numProbes++
+				if apiServerAvailable {
+					updateSuccessMetrics(a.apiAvailabilityMetrics)
+				} else {
+					updateFailureMetrics(a.apiAvailabilityMetrics)
+				}
+				return
+			}
+
+		}
+	}()
+	return nil, nil
+}
+
+func updateFailureMetrics(apiAvailabilityMetrics *apiAvailabilityMetrics) {
+	// increment total number of failed probes count, current number of failed probes by 1
+	apiAvailabilityMetrics.numFailures++
+	apiAvailabilityMetrics.consecutiveFailedProbes++
+	// if the previous probe was not a failure, then set the previous probe failure flag to true.
+	// This would be used by next probe
+	if !apiAvailabilityMetrics.prevProbeFailure {
+		apiAvailabilityMetrics.prevProbeFailure = true
```

Comment:
Why do we need it? if consecutiveFailedProbes == 0 (or 1 at this poin in the code) that means that prevProbe wasn't a failure, otherwise it was).

Reply:
Well, I hadn't thought of doing it via the consecutiveFailedProbes counter. On second thoughts, when I re-looked at the code, the checks with a boolean flag seemed a bit more intuitive to understand that the previous probe was a failure than a counter based check. That's my personal opinion though and can change it if it seems complicated.

Reply:
The less variables/fields we have the better.

If you want, you can introduce a method, like:
```
func (m *apiAvailabilitymetrics) previousProbeFailed() bool {
}
```

that will return what you want and use it where you want.
Let's not introduce unnecessary vars that you have to keep in sync.

Reply:
I made that change.